### ISSUE TITLE
Entries of manifest out of range causes direct mode miss

### ISSUE THREAD

**Issue by MizuhaHimuraki** [ccache/ccache#493]
First, I execute a build with a.cpp which include a header named pub.h, ccache created cache and manifest correctly.

Then, I define a macro in the public header  pub.h
> #define A_CPP_NOT_USE "1"

When I build again, in the compiling of a.cpp, the manifest will not hit cache in direct mode since the change of size, but ccache will hit the cache in preprocessor mode, and overwrite the old manifest.

Later, I rollback pub.h, and built one more time, it missed cache in direct mode again, and rollbacked manifest.

Should we put the multiple possible information in manifest instead of simply overwriting?

We use ccache with distributed nfs system in a giant C++ project development, developers may run pretest base on different commit or branch, manifest repeatedly overwrited again and again.

I built the binary with same commit, and I got the following hit rate, I think those all could be hitted in direct mode

> cache hit (direct)                  2150
cache hit (preprocessed)              81
cache miss                             1
cache hit rate                     99.96 %
cleanups performed                     0
files in cache                         1
cache size                         380.9 kB

Maybe we can use an option like "CCACHE_OVERWRITE_MANIFEST" to control this feature?

Looking forward to your reply

**Comment by jrosdahl** [comment]
> First, I execute a build with a.cpp which include a header named pub.h, ccache created cache and manifest correctly.
> 
> Then, I define a macro in the public header pub.h
> 
> > #define A_CPP_NOT_USE "1"
> 
> When I build again, in the compiling of a.cpp, the manifest will not hit cache in direct mode since the change of size, but ccache will hit the cache in preprocessor mode,

Sounds expected so far.

> and overwrite the old manifest.

No, that's not what happens (unless there's some bug, of course). A manifest stores multiple entries and old entries are not overwritten.

It should be noted, though, that manifest files are intentionally not protected via file locks, so what can happen is that some entries may be lost if several compilations want to update same manifest file _at the same time_. The latest update will win and the manifest file won't be corrupted, only potentially missing the entry from the "loser compilation".

> Later, I rollback pub.h, and built one more time, it missed cache in direct mode again

That's unexpected. I cannot reproduce this. Here's a script you can test:

```sh
#!/bin/sh

ccache=ccache
cc=/usr/bin/gcc

echo '#include "pub.h"' >main.c
echo '' >pub.h
sleep 2 # Need to sleep in order to avoid invalidation of a too new header file
CCACHE_RECACHE=1 CCACHE_LOGFILE=/dev/stdout $ccache $cc -c main.c | grep Result
CCACHE_LOGFILE=/dev/stdout $ccache $cc -c main.c | grep Result

echo '#define A_CPP_NOT_USE "1"' >pub.h
sleep 2
CCACHE_LOGFILE=/dev/stdout $ccache $cc -c main.c | grep Result

echo '' >pub.h
sleep 2
CCACHE_LOGFILE=/dev/stdout $ccache $cc -c main.c | grep Result
```
Output for me first time running the script using ccache 3.7.6:

```
[2019-11-27T20:50:46.866016 30344] Result: cache miss
[2019-11-27T20:50:46.867502 30352] Result: cache hit (direct)
[2019-11-27T20:50:48.878694 30390] Result: cache hit (preprocessed)
[2019-11-27T20:50:50.883255 30398] Result: cache hit (direct)
```

Second time running the script:

```
[2019-11-27T20:51:00.563311 30886] Result: cache miss
[2019-11-27T20:51:00.565315 30894] Result: cache hit (direct)
[2019-11-27T20:51:02.568085 30908] Result: cache hit (direct)
[2019-11-27T20:51:04.570871 30988] Result: cache hit (direct)
```

Try running the script on the NFS mount and on a local mount to see if there's any difference.

> Should we put the multiple possible information in manifest instead of simply overwriting?

As described above, that is already done.

> We use ccache with distributed nfs system in a giant C++ project development, developers may run pretest base on different commit or branch, manifest repeatedly overwrited again and again.

Perhaps the behavior you see could be an artefact of your using NFS? ccache relies on atomicity of file operations and NFS is notoriously flaky.

**Comment by MizuhaHimuraki** [comment]
Thank you jrosdahl, logs in https://github.com/ccache/ccache/pull/494 show my header file is overwrited in manfiest, I thought this was a feature of ccache, maybe I should try to solve this problem on NFS

**Comment by MizuhaHimuraki** [comment]
> Perhaps the behavior you see could be an artefact of your using NFS? ccache relies on atomicity of file operations and NFS is notoriously flaky.

Hi jrosdahl, I read the code of manifest_put and found that it is not fully as you described.
https://github.com/ccache/ccache/blob/1a4929bfd73e63ebd1b369ee116f9c57f1931824/src/manifest.c#L650

See here, manifest_put is not an atomic operation, no write lock between read and rewrite, it allows for the possibility of overwriting. 

Our ci system has at most 7 build task and over 600 cores to compile with one cache_dir at the same time, in this case, both NFS and local disks have a high probability of being overwritted.

I think, this is why you cannot reproduce it in a single-threaded case.

**Comment by MizuhaHimuraki** [comment]
But this still doesn't explain why the hit rate of the same code drops in direct mode.  Even if there is a possible overwrite, it will only cause the incremental data to write to disk failed. Absolutely, it will not affect the hit rate. I will try to find a more reasonable explanation.

**Comment by MizuhaHimuraki** [comment]
https://github.com/ccache/ccache/blob/1a4929bfd73e63ebd1b369ee116f9c57f1931824/src/manifest.c#L677-L700

I found a feature of manifest not mentioned in the documentation, if object entries or file_info more than 10000, ccache will discard old manfiest, and create a new one.

And I really find the key words in my recent build logs.

> ccache.log:[2019-11-28T15:14:56.476173 70396] More than 10000 file_info entries in manifest file; discarding.

Maybe ccahce can use some better ways, such as LRU algorithm, to optimize the discarding policy of manifest objects or use stat tables to support hit cache in larger manifests in future.

Anyway, I will try to compile ccache with larger "MAX_MANIFEST_ENTRIES" first. : )

**Comment by MizuhaHimuraki** [comment]
I think my problem is clear now. Should I open another two issue to talk about the following features?

#### Use config items to control the limit of manifest
- CCACHE_MAX_MANIFEST_ENTRIES
- CCACHE_MAX_MANIFEST_FILE_INFO_ENTRIES

#### Add a new sloppiness item named "file_stat_no_time"
- this option conflicts with 'file_stat_matches' and 'file_stat_matches_ctime'
- all mtime and ctime in manifest is set to 0
I think, this feature may be useful in ci for a large project, since the mtime and ctime may be edited by git again and again
I dumped one of my manifest, the output shows the manifest owns 13 objects, and 720 headers, but the num of file infos is more than 8K, lots of header files owns the same hash and size, the only different is their mtime and ctime.

**Comment by jrosdahl** [comment]
> I abstracted the problem I encountered too simply before. 

Agreed. :slightly_smiling_face:

> I found a feature of manifest not mentioned in the documentation, if object entries or file_info more than 10000, ccache will discard old manfiest, and create a new one.
> [...]
> And I really find the key words in my recent build logs.

Great that you found the reason!

> Maybe ccahce can use some better ways, such as LRU algorithm, to optimize the discarding policy of manifest objects or use stat tables to support hit cache in larger manifests in future.

For clarification: Even though the comment for the manifest entries check talks about LRU, doing eviction based on LRU or other methods would imply writing data for direct mode hits, which is nice to avoid if possible.

> Should I open another two issue to talk about the following features?

Let's continue here for a bit longer.

So if I understand correctly, the problematic scenario is this:

1. A CI system checks out source code files which get new mtimes and ctimes each time.
2. When compiling a previously compiled (unmodified) source code file X.c which references at least one include file that has changed content, there will be a direct mode miss.
3. The direct mode miss will result in adding new file info entries to the manifest for **all** (local) header files referenced by X.c since all (local) header files have new mtime/ctime values.
4. If X.c references many include files, the file info limit can quickly be reached, which will result in the manifest file being thrown away.

I don't think that the solution to this is to make the maximum number of results and file infos configurable since that will only result in collecting even more unwanted data in the manifest. Instead we should make sure that unwanted data entries are not added in the first place. This is just like your file_stat_no_time idea, except that I don't think that it needs to be configurable.

ccache currently always stores mtime/ctime information but only reads it if sloppy_file_stat is set. This doesn't really make sense – the mtime/ctime information should only be added when expected to be read. So my suggestion it to only store mtimes/ctimes when sloppy_file_stat is set and store them as zeroes otherwise (using sloppy_file_stat only makes sense to set when compiling on a single system, not in a CI situation). Sounds reasonable?

**Comment by MizuhaHimuraki** [comment]
> I don't think that the solution to this is to make the maximum number of results and file infos configurable since that will only result in collecting even more unwanted data in the manifest. Instead we should make sure that unwanted data entries are not added in the first place. This is just like your file_stat_no_time idea, except that I don't think that it needs to be configurable.

I think this limit should be configurable. I agree with some considerations based on performance, but you could not prove that 10K is very reasonable in any scenario. 
However, before there is a better solution, we can add some descriptions of this feature to the document first.

> ccache currently always stores mtime/ctime information but only reads it if sloppy_file_stat is set. This doesn't really make sense – the mtime/ctime information should only be added when expected to be read. So my suggestion it to only store mtimes/ctimes when sloppy_file_stat is set and store them as zeroes otherwise (using sloppy_file_stat only makes sense to set when compiling on a single system, not in a CI situation). Sounds reasonable?

Obviously your solution is better, I can't agree more! And I created a PR above.

**Comment by jrosdahl** [comment]
> I think this limit should be configurable.

How many people do you think would understand what such a configuration setting does and then would like to configure it? In this case I think that it's better to find good limits and be done with it so that it Just Works. The limits are intended to be high enough for all common scenarios. I would prefer to avoid adding configuration options for small details like this so that the number of configuration options doesn't grow too much.

> [...] you could not prove that 10K is very reasonable in any scenario.

I don't quite understand what you mean by this. What is there to prove?

> However, before there is a better solution, we can add some descriptions of this feature to the document first.

I don't see it as a feature but more as an internal safety measure that should not be triggered unless something goes south. It would certainly be possible to add something about it to the manual (I guess that's what you meant with document), but since the manual doesn't describe the manifest file in detail I don't think that there currently is a good place to put the information.

**Comment by MizuhaHimuraki** [comment]
Well, your statement makes more sense in current state. I think I have encountered this problem very early when I tried #438, I thought it was a bug in that fork.

> you could not prove that 10K is very reasonable in any scenario.
> I don't see it as a feature but more as an internal safety measure that should not be triggered unless something goes south.

I think ccache is mostly used to speed up personal builds, but if you use ccache to create a cloud build service (I think my  distributed nfs system with distcc is one of the instances), this will be a problem.

Maybe this topic will be discussed more when others got the same issue after the backend API is provided.

Anyway, my problem is solved by #495, It works well with the 10K limit.
If a got something new, I will open another issue to talk about the manifest.

Thanks a lot.😀
